File: LRUCache.py
class Node:
    def __init__(self, key=0, value=0):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = Node()  # dummy head
        self.tail = Node()  # dummy tail
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove_node(self, node):
        prev = node.prev
        next = node.next
        prev.next = next
        next.prev = prev

    def _add_to_tail(self, node):
        prev = self.tail.prev
        prev.next = node
        node.prev = prev
        node.next = self.tail
        self.tail.prev = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._remove_node(node)
        self._add_to_tail(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self._remove_node(node)
            self._add_to_tail(node)
        else:
            if len(self.cache) >= self.capacity:
                lru = self.head.next
                self._remove_node(lru)
                del self.cache[lru.key]
            new_node = Node(key, value)
            self.cache[key] = new_node
            self._add_to_tail(new_node)

lru_cache = LRUCache(2)
lru_cache.put(1,1)
lru_cache.put(2,2)
print(lru_cache.get(1))
lru_cache.put(3,3)
print(lru_cache.get(2))
lru_cache.put(4,4)
print(lru_cache.get(1))
print(lru_cache.get(3))
print(lru_cache.get(4))

        

File: LRUCache.java
import java.util.HashMap;
import java.util.Map;

public class LRUCache{
    private class Node{
        private int key;
        private int value;
        private Node prev;
        private Node next;
        public Node(int key, int value){
            this.key = key;
            this.value = value;
            this.prev = null;
            this.next  = null;
        }
    }

    private int capacity;
    private Map<Integer, Node> cache;
    private LRUCache.Node start;
    private LRUCache.Node end;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.cache = new HashMap<>(capacity);
        start = new LRUCache.Node(0, 0);
        end = new LRUCache.Node(0, 0);
        start.next = end;
        end.prev = start;

    }

    private void appendAtEnd(Node node){
        var currentLast = end.prev;
        currentLast.next = node;
        node.prev = currentLast;
        node.next = end;
        end.prev = node;
    }

    private void removeNode(Node node){
        var prevNode = node.prev;
        var nextNode = node.next;
        prevNode.next = nextNode;
        nextNode.prev = prevNode;
    }
    
    public int get(int key) {
        if(!cache.containsKey(key)){
            return -1;
        }
        Node currentNode = cache.get(key);
        if(currentNode != end.prev){
            removeNode(currentNode);
            appendAtEnd(currentNode);
        }
        return currentNode.value;
    }
    
    public void put(int key, int value) {
        if(cache.containsKey(key)){
            var node = cache.get(key);
            removeNode(node);
            appendAtEnd(node);
            node.value = value;
            return;
        }
        if(capacity <= cache.size()){
            var removeNode = start.next;
            cache.remove(removeNode.key);
            removeNode(removeNode);
        }
        Node node = new Node(key, value);
        appendAtEnd(node);
        cache.put(key, node);
        
    }

    public static void main(String[] args) {
        LRUCache lruCache = new LRUCache(2);
        System.out.println(lruCache.get(2));
        lruCache.put(2,6);
        System.out.println(lruCache.get(1));
        lruCache.put(1, 5);
        lruCache.put(1, 2);
        System.out.println(lruCache.get(1));
        System.out.println(lruCache.get(2));
    }
}